### PR TITLE
02_As a user, I need to preview an announcement I just created

### PR DESCRIPTION
**Pull request by Nicofuma**
Issue: #6

### CODE REVIEW

File: app/config/services.yml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+services:
+    app.form.type.job:
+        class: SensioLabs\JobBoardBundle\Form\Type\JobType
+        tags:
+            - { name: form.type, alias: job }
+
+    app.twig.extension:
```
Comment (salahm) [inline app/config/services.yml]: i think, this is out of scope too. the us is about only previewing an announcement

File: src/SensioLabs/JobBoardBundle/Controller/BaseController.php [hunk 2]
```diff
@@ -28,7 +30,24 @@ public function indexAction(Request $request)
...
+
+        if ($form->isValid()) {
+            $em = $this->getDoctrine()->getManager();
+            $em->persist($job);
+            $em->flush();
+
+            return $this->redirectToRoute('job_preview', [
+                'country'  => $job->getCountry(),
+                'contract' => $job->getContractType(),
+                'slug'     => $job->getSlug()]);
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Controller/BaseController.php]: weird CS here (`])`)

File: src/SensioLabs/JobBoardBundle/Entity/Job.php [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+     *
+     * @Assert\Length(max=255)
+     * @Assert\NotBlank(message="Job title should not be empty")
+     */
+    private $title;
+
+    /**
+     * @ORM\Column(name="country", type="string", length=255)
+     *
+     * @Assert\Length(max=255)
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: IMO Length constraint is not needed here. Country do the needed job

Comment (Nicofuma) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: Updated in the previous PR and rebased

File: src/SensioLabs/JobBoardBundle/Twig/JobBoardExtension.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    /**
+     * Replace a contract type code by it's pretty name
+     *
+     * @param string $contractType
+     *
+     * @return null|string
+     */
+    public function filterContractType($contractType)
+    {
+           $contractTypes = Job::getContractTypes();
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Twig/JobBoardExtension.php]: wow such an indentation

File: app/config/config.yml [hunk 5]
```diff
@@ -53,7 +54,7 @@ doctrine:
         charset:  UTF8
         # if using pdo_sqlite as your database driver, add the path in parameters.yml
         # e.g. database_path: %kernel.root_dir%/data/data.db3
-        # path:     %database_path%
+        path:     %database_path%
```
Comment (florianbelhomme) [inline app/config/config.yml]: SQLite is used in every env ?

File: app/config/services.yml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+services:
+    app.form.type.job:
```
Comment (florianbelhomme) [inline app/config/services.yml]: The services must be declared in the bundle JobBoardBundle and should be named sensiolabs.jobboard.my.service

File: composer.json [hunk 7]
```diff
@@ -12,7 +12,7 @@
         }
     ],
     "require": {
-        "php": ">=5.4",
+        "php": ">=5.5",
```
Comment (florianbelhomme) [inline composer.json]: What did you add to need PHP 5.5 ? Have you seen the client to know if it's OK ?

File: src/SensioLabs/JobBoardBundle/Controller/JobController.php [hunk 8]
```diff
@@ -19,19 +21,27 @@ public function showAction()
     }
 
     /**
-     * @Route("/preview", name="job_preview")
+     * @Route(
+     *   name="job_preview",
+     *   path="/{country}/{contract}/{slug}/preview"
+     * )
+     * @ParamConverter("job", options={"mapping"={"country"="country","contract"="contractType","slug"="slug"}})
```
Comment (florianbelhomme) [inline src/SensioLabs/JobBoardBundle/Controller/JobController.php]: Jobs can share the same slug ? Do you need to match the country and the contract ?

Comment (Nicofuma) [inline src/SensioLabs/JobBoardBundle/Controller/JobController.php]: It's not needed but it's preferable to validate them. Otherwise this URL would be valid: `/they/suck/sensiolabs-developer-senior-1/preview`

File: composer.json (hunk 7 above)
Comment (Nicofuma) [inline composer.json]: it has been discussed yes

File: app/config/config.yml (hunk 5 above)
Comment (Nicofuma) [inline app/config/config.yml]: in my setup, yes

File: src/SensioLabs/JobBoardBundle/Controller/JobController.php (hunk 8 above)
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Controller/JobController.php]: :+1:

File: composer.json (hunk 7 above)
Comment (ogizanagi) [inline composer.json]: What are the arguments about the php 5.5 limitation ? Why not 5.6 which will still be actively supported in the next eleven months, whereas 5.5 is solely supported for security fixes ?

Comment (Nicofuma) [inline composer.json]: bumped to 5.6 actually

File: src/SensioLabs/JobBoardBundle/Entity/Job.php [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+     */
+    private $company;
+
+    /**
+     * @Gedmo\Slug(fields={"title"})
+     * @ORM\Column(length=255, unique=true)
+     */
+    private $slug;
+
+    public static function getContractTypesValues()
```
Comment (salahm) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: well the name is weird, get*Values with and an `array_keys()`, to get "machine names" ? :volcano:

File: src/SensioLabs/JobBoardBundle/Entity/Job.php [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+    /**
+     * @ORM\Column(name="title", type="string", length=255)
+     *
+     * @Assert\Length(max=255)
+     * @Assert\NotBlank(message="Job title should not be empty")
+     */
+    private $title;
+
+    /**
+     * @ORM\Column(name="country", type="string", length=255)
```
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: length 255 ?

File: src/SensioLabs/JobBoardBundle/Entity/Job.php (hunk 9 above)
Comment (ogizanagi) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: I'm used to have two simple `getContractTypes` and `getReadableContractTypes` methods personally, where the readable ones could be the translation key if supported.

Comment (Nicofuma) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: I like @ogizanagi solution, pushed.

File: src/SensioLabs/JobBoardBundle/Entity/Job.php (hunk 10 above)
Comment (Nicofuma) [inline src/SensioLabs/JobBoardBundle/Entity/Job.php]: fixed (changed to 2)

**Comment by ogizanagi** [comment]
:+1:
